### PR TITLE
[KBFS-2756] Show Files in Finder

### PR DESCRIPTION
**Pull request by jzila**
This PR implements the per-row widget for showing files in Finder. Currently it only shows up if you hover over the correct spot on the row. In the future I want it to show up if you hover over the whole row, but it seems like that will take plumbing through a bunch of state.

### CODE REVIEW

File: shared/actions/fs-platform-specific.desktop.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    // For example /keybase/private/gabrielh,chris gets redirected to
+    // /keybase/private/chris,gabrielh.
+    fs.realpath(openPath, (err, resolvedPath) => {
+      if (err) {
+        reject(new Error(`No realpath for ${openPath}: ${err}`))
+        return
+      }
+      // Convert to URL for openExternal call.
+      // We use openExternal instead of openItem because it
+      // correctly focuses' the Finder, and also uses a newer
```
Comment (songgao) [inline shared/actions/fs-platform-specific.desktop.js]: I know this is from the original kbfs saga, but I don't quite understand this part. When I experimented with `shell.openExternal`, it doesn't focus items in Finder, but would just open it (like when you use the `open` command).

File: shared/actions/fs-platform-specific.desktop.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      }
+    })
+  })
+}
+
+function _open(openPath: string): Promise<*> {
+  return new Promise((resolve, reject) => {
+    isDirectory(openPath).then(isDir => {
+      if (isDir && isWindows) {
+        if (!shell.openItem(openPath)) {
```
Comment (songgao) [inline shared/actions/fs-platform-specific.desktop.js]: Should windows be using the same code path?

File: shared/fs/index.js [hunk 3]
```diff
@@ -102,6 +98,28 @@ class Files extends React.PureComponent<FolderProps> {
   }
 }
 
+const folderBoxStyle = {...globalStyles.flexBoxColumn, flex: 1, justifyContent: 'space-between'}
```
Comment (songgao) [inline shared/fs/index.js]: Are we having a convention here to always have style definitions after the components?

File: shared/actions/fs-platform-specific.desktop.js (hunk 2 above)
Comment (jzila) [reply shared/actions/fs-platform-specific.desktop.js]: Not sure, that's why I was asking earlier about the `openInCurrentMountDir` generator in `actions/kbfs.desktop.js`.

Comment (jzila) [reply shared/actions/fs-platform-specific.desktop.js]: I'm going to test it this weekend.

File: shared/actions/fs-platform-specific.desktop.js (hunk 1 above)
Comment (jzila) [reply shared/actions/fs-platform-specific.desktop.js]: You're right, it doesn't seem to function according to the comment, at least not on Linux. I'll test macOS and Windows as well.

Comment (songgao) [reply shared/actions/fs-platform-specific.desktop.js]: Yeah, I think the showItemInFolder is the only API that does it: https://github.com/electron/electron/blob/master/docs/api/shell.md

File: shared/actions/fs-platform-specific.desktop.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+function isDirectory(openPath: string): Promise<boolean> {
+  return new Promise((resolve, reject) => {
+    fs.stat(openPath, (err, stats) => {
+      if (err) {
+        reject(new Error(`Unable to open/stat file: ${openPath}`))
+        return
+      }
+      if (stats.isFile()) {
```
Comment (chrisnojima) [inline shared/actions/fs-platform-specific.desktop.js]: why call isFile?

File: shared/actions/fs-platform-specific.desktop.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      } else {
+        reject(new Error(`Unable to open: Not a file or directory`))
+      }
+    })
+  })
+}
+
+function _open(openPath: string): Promise<*> {
+  return new Promise((resolve, reject) => {
+    isDirectory(openPath).then(isDir => {
```
Comment (chrisnojima) [inline shared/actions/fs-platform-specific.desktop.js]: do all paths actually resolve or reject, doesn't seem like it to me. isDirectory can throw and it won't plumb that through

File: shared/actions/kbfs.desktop.js [hunk 6]
```diff
@@ -297,32 +295,7 @@ function* openWithCurrentMountDir(openPath: string): Saga.SagaGenerator<any, any
   yield Saga.call(_open, resolvedPath)
 }
 
-function* openSaga(action: KBFSGen.OpenPayload): Saga.SagaGenerator<any, any> {
```
Comment (chrisnojima) [inline shared/actions/kbfs.desktop.js]: i'd advise leaving the old actions alone or you'll start breaking the app in production

File: shared/chat/conversation/attachment-popup/container.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // @flow
 import * as Types from '../../../constants/types/chat'
 import * as ChatGen from '../../../actions/chat-gen'
-import * as KBFSGen from '../../../actions/kbfs-gen'
```
Comment (chrisnojima) [inline shared/chat/conversation/attachment-popup/container.js]: why change this the existing code for this? why not just leave it

Comment (jzila) [reply shared/chat/conversation/attachment-popup/container.js]: To start migrating to the new stuff.

Comment (chrisnojima) [reply shared/chat/conversation/attachment-popup/container.js]: we should chat

File: shared/actions/kbfs.desktop.js (hunk 6 above)
Comment (jzila) [reply shared/actions/kbfs.desktop.js]: Right now the code is basically a straight port. I'm hoping that we can consolidate to the new code and fix all the bugs in one place.

File: shared/actions/fs-platform-specific.desktop.js (hunk 4 above)
Comment (jzila) [reply shared/actions/fs-platform-specific.desktop.js]: copy-paste. I'll remove it.

**Comment by jzila** [comment]
@chrisnojima: ready for another look

File: shared/fs/row.js [hunk 8]
```diff
@@ -24,13 +25,14 @@ const mapDispatchToProps = (dispatch: Dispatch) => ({
     if (type === 'folder') {
       dispatch(navigateAppend([{props: {path}, selected: 'folder'}]))
     } else {
-      dispatch(FsGen.createDownload({path}))
+      dispatch(FSGen.createDownload({path}))
       console.log('Cannot view files yet. Requested file: ' + Types.pathToString(path))
     }
   },
+  _openInFileUI: (path: Types.Path) => dispatch(FSGen.createOpenInFileUI({path: Types.pathToString(path)})),
```
Comment (songgao) [inline shared/fs/row.js]: After rebasing to this branch, I'm realizing that this is a plain "open" call, which I guess is why you filtered on `type === folder` in the component. It seems a "show in Finder" button probably makes more sense, especially for files?

Comment (jzila) [reply shared/fs/row.js]: I don't follow...for folders, we open the target path. For files, we `shell.showItemInFolder`.

Comment (songgao) [reply shared/fs/row.js]: > we open the target path. For files, we shell.showItemInFolder.

That sounds good with me. (I feel like we talked about this in a meeting but couldn't remember clearly.) Currently in this PR we only show icon for folders though. But I'm OK with adding it later too; just thought I'd bring it up.

Comment (jzila) [reply shared/fs/row.js]: Ah yeah. I didn't want to step on your toes re: clicking files. We can add it once your downloading code is merged.

**Comment by chrisnojima** [comment]
i belive the intention of that action is to navigate you to the root of the folder tab and switch the tab there. This is encodable as one route action (but i think its actually pretty confusing, i have to keep looking it up)
The sequentially helper just does actions one at a time

**Comment by jzila** [comment]
@chrisnojima why isn't it just a single `navigateTo`?

**Comment by songgao** [comment]
Sorry for hijacking the discussion, but it'd be nice to be able to thumb/merge this so we can rebase other PRs on top of master :)

**Comment by chrisnojima** [comment]
yes it could be is what i'm saying

File: shared/actions/fs-platform-specific.desktop.js (hunk 4 above)
Comment (chrisnojima) [reply shared/actions/fs-platform-specific.desktop.js]: still there

Comment (jzila) [reply shared/actions/fs-platform-specific.desktop.js]: Yes but now it makes sense :P

Comment (jzila) [reply shared/actions/fs-platform-specific.desktop.js]: Basically this method wasn't `isDirectory`, it was "is this a file or a directory, and if so, which one?"

**Pull request merged**